rfc/src/rfcs/0004-loop-contract-synthesis.md

```diff
@@ -0,0 +1,145 @@
+        `inner`: while x > 0 {
+            y = y - 1;
+            x = x - 1;
+        }
+        x = 2;
+    }
+    assert!(y == 0 || y == -1);
+}
+```
+Then, we the flag `--synthesize-loop-contracts` is present, the synthesizer will synthesize loop contracts for only the loop `outer` in the function `bar` and unwind the other three loops.
+
+### Limit Resource Used by Synthesizer for Termination
+Without resource limit, an enumerative synthesizer may run forever to exhaust search space consist of infinite number of candidates, especially when there is no solution in the search space. So, for the guarantee of termination, we provide users options: <code>--limit-synthesis-time <ins>T</ins></code> to limit the running time of the synthesizer, and <code>--limit-synthesis-size <ins>N</ins></code> to limit the sizes (number of nodes in the AST of the candidate predicate) of loop-invariant candidates we search for during the enumeration.
+
+We encourage users to specify such limit while we hardcode some default limit at the same time. 
+
+### Output of Kani when Synthesizer Enable
+When the flag `--synthesize-loop-contracts` is provided, Kani will report different result for different cases
+1. When there exists some loop invariant in the candidate space with which all assertions can be proved, Kani will synthesize the loop contracts, verify the program with the synthesized loop contracts, and report verification SUCCESS;
+2. When there exists no loop invariants in the candidate space that can be used to prove all assertions within the resource limit, Kani will report the verification result with the best-effort-synthesized loop contracts.
+
+A question about how do we print the synthesized loop contracts when users request is discussed in **Open question**.
+
+## Detailed Design
+The synthesizer ```goto-synthesizer``` is implemented in the repository of `CBMC`, takes as input a goto binary, and outputs a new goto binary with the synthesized loop contracts applied. Currently, Kani invoke `goto-instrument` to instrument the goto binary `main.goto` into a new goto binary `main_instrumented.goto`, and then invoke ```cbmc``` on `main_instrumented.goto` to get the verification result. The synthesis will happen between calling `goto-instrument` and calling ```cbmc```. That is, we invoke ```goto-synthesizer``` on ```main_instrumented.goto``` to produce a new goto binary ```main_synthesized.goto```, and then call ```cbmc``` on `main_synthesized.goto` instead. 
+
+When invoking ```goto-synthesizer```, we pass the following parameters to it with the flags built in `goto-synthesizer`:
+- the resource limit of the synthesis;
+- the unwinding numbers of loops.
+Note that the way we identify loops in goto and the way we identify loop in rust are different. So we also need to parse the label-based loop id to the index-based loop id used by goto.
+
+
+The enumerator used in the synthesizer enumerate candidates from the language of the following grammar template
```

Comment:
I don't really understand what this means, I agree with @feliperodri comment about explaining about candidates.

I wonder, would it be practical to give an example here? I.e. for a particular snippet of code, show what the resulting candidate would be? It's difficult for me to relate these constructs to what I normally think of as an assertion